# PHASE 10 — FINAL BANKING RISK & CROSS-SELL SYSTEM

## Objective

The objective of Phase 10 is to integrate the final credit-risk prediction
model developed in the previous phases with the cross-sell recommendation
system developed using association rules.

The final system will:

1. Load the trained credit-risk model.
2. Load the final feature list.
3. Load customer risk data.
4. Predict the probability of credit-card default.
5. Classify customers based on the selected threshold.
6. Generate a banking risk decision.
7. Integrate cross-sell product recommendations.
8. Generate a final customer-level analysis.
9. Provide an interactive banking decision-support system.

### Final System Flow

Customer Data
      ↓
Feature Preparation
      ↓
Random Forest Model
      ↓
Default Probability
      ↓
Risk Classification
      ↓
Banking Decision
      ↓
Cross-Sell Recommendation
      ↓
Final Customer Analysis

In [1]:
import pandas as pd
import numpy as np
import joblib
import os

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
print("Current working directory:")
print(os.getcwd())

print("\nFiles available:")

for file in os.listdir():
    print(file)

Current working directory:
c:\Users\Meghana\OneDrive\Desktop\BANKING-CREDIT-RISK\notebooks

Files available:
01_dataset_check.ipynb
01_eda.ipynb
02_preprocessing.ipynb
03_model_training.ipynb
04_association_rules.ipynb
05_final_banking_system.ipynb
credit_default_feature_engineered.csv
customer_cross_sell_recommendations.csv
customer_products.csv
customer_products_v2.csv
final_credit_risk_model.pkl
final_features.pkl


In [3]:
model = joblib.load("final_credit_risk_model.pkl")

print("Final Random Forest model loaded successfully!")

Final Random Forest model loaded successfully!


In [4]:
selected_features = joblib.load("final_features.pkl")

print("Final feature list loaded successfully!")
print("Number of features:", len(selected_features))
print(selected_features)

Final feature list loaded successfully!
Number of features: 23
['LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_SEPT', 'PAY_AUG', 'PAY_JUL', 'PAY_JUN', 'PAY_MAY', 'PAY_APR', 'BILL_AMT_SEPT', 'BILL_AMT_AUG', 'BILL_AMT_JUL', 'BILL_AMT_JUN', 'BILL_AMT_MAY', 'BILL_AMT_APR', 'PAY_AMT_SEPT', 'PAY_AMT_AUG', 'PAY_AMT_JUL', 'PAY_AMT_JUN', 'PAY_AMT_MAY', 'PAY_AMT_APR']


In [5]:
FINAL_THRESHOLD = 0.25

print("Final Risk Threshold:", FINAL_THRESHOLD)

Final Risk Threshold: 0.25


In [6]:
cross_sell_df = pd.read_csv("customer_cross_sell_recommendations.csv")

print("Cross-sell dataset loaded successfully!")
print("Shape:", cross_sell_df.shape)
print("\nColumns:")
print(cross_sell_df.columns.tolist())

Cross-sell dataset loaded successfully!
Shape: (22190, 6)

Columns:
['Customer_ID', 'Existing_Products', 'Recommended_Product', 'Confidence', 'Lift', 'Support']


In [7]:
display(cross_sell_df.head())

,Customer_ID,Existing_Products,Recommended_Product,Confidence,Lift,Support
0,1,"Savings_Account, Debit_Card",Credit_Card,0.601543,1.267487,0.414715
1,3,Debit_Card,Credit_Card,0.538636,1.134939,0.431615
2,3,Debit_Card,Savings_Account,0.860364,1.121618,0.689419
3,9,"Savings_Account, Insurance",Credit_Card,0.556027,1.171583,0.162302
4,9,"Savings_Account, Insurance",Debit_Card,0.898765,1.121618,0.689419


In [8]:
customer_products = pd.read_csv("customer_products.csv")
customer_products_v2 = pd.read_csv("customer_products_v2.csv")

print("customer_products")
print("Shape:", customer_products.shape)
print("Columns:")
print(customer_products.columns.tolist())

print("\n" + "="*60)

print("customer_products_v2")
print("Shape:", customer_products_v2.shape)
print("Columns:")
print(customer_products_v2.columns.tolist())

customer_products
Shape: (29965, 7)
Columns:
['Customer_ID', 'Savings_Account', 'Credit_Card', 'Debit_Card', 'Insurance', 'Home_Loan', 'Personal_Loan']

customer_products_v2
Shape: (29965, 7)
Columns:
['Customer_ID', 'Savings_Account', 'Debit_Card', 'Home_Loan', 'Personal_Loan', 'Insurance', 'Credit_Card']


In [9]:
print("\n--- customer_products ---")
display(customer_products.head(10))

print("\n--- customer_products_v2 ---")
display(customer_products_v2.head(10))


--- customer_products ---


,Customer_ID,Savings_Account,Credit_Card,Debit_Card,Insurance,Home_Loan,Personal_Loan
0,1,1,0,1,0,0,1
1,2,0,0,0,0,0,0
2,3,0,1,1,0,0,0
3,4,1,0,1,1,0,0
4,5,1,0,1,0,0,0
5,6,1,0,1,1,0,0
6,7,1,1,1,0,0,0
7,8,0,0,1,0,0,0
8,9,1,1,1,0,1,0
9,10,0,0,1,0,0,0



--- customer_products_v2 ---


,Customer_ID,Savings_Account,Debit_Card,Home_Loan,Personal_Loan,Insurance,Credit_Card
0,1,1,1,0,0,0,0
1,2,0,0,0,0,0,0
2,3,0,1,0,0,0,0
3,4,1,1,0,1,1,1
4,5,1,1,0,0,1,1
5,6,1,1,0,0,0,1
6,7,1,1,0,0,0,1
7,8,0,0,0,0,0,0
8,9,1,0,0,0,1,0
9,10,0,0,0,0,0,0


In [11]:
print("customer_products Customer_ID:")
print(customer_products["Customer_ID"].head(10).tolist())

print("\ncustomer_products_v2 Customer_ID:")
print(customer_products_v2["Customer_ID"].head(10).tolist())

print("\nCross-sell Customer_ID:")
print(cross_sell_df["Customer_ID"].head(10).tolist())

customer_products Customer_ID:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

customer_products_v2 Customer_ID:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Cross-sell Customer_ID:
[1, 3, 3, 9, 9, 13, 17, 17, 18, 20]


In [13]:
risk_df = pd.read_csv("credit_default_feature_engineered.csv")

print("Risk dataset loaded successfully!")
print("Shape:", risk_df.shape)

print("\nNumber of rows:", len(risk_df))
print("Number of columns:", len(risk_df.columns))

print("\nFirst 5 rows:")
display(risk_df.head())

Risk dataset loaded successfully!
Shape: (29965, 45)

Number of rows: 29965
Number of columns: 45

First 5 rows:


,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_SEPT,PAY_AUG,PAY_JUL,PAY_JUN,PAY_MAY,...,Severe_Delay_Count,Recent_Payment_Delay,Recent_Avg_Delay,Older_Avg_Delay,Payment_Delay_Trend,Age_Category,Credit_Limit_Category,High_Utilization_Flag,Frequent_Delay_Flag,Severe_Delay_Flag
0,20000,2,2,1,24,2,2,-1,-1,-2,...,2,2,1.000000,-1.666667,2.666667,Young,Low,0,0,1
1,120000,2,2,2,26,-1,2,0,0,0,...,2,-1,0.333333,0.666667,-0.333333,Adult,Medium,0,0,1
2,90000,2,2,2,34,0,0,0,0,0,...,0,0,0.000000,0.000000,0.000000,Adult,Medium,0,0,0
3,50000,2,2,1,37,0,0,0,0,0,...,0,0,0.000000,0.000000,0.000000,Adult,Low,1,0,0
4,50000,1,2,1,57,-1,0,-1,0,0,...,0,-1,-0.666667,0.000000,-0.666667,Middle_Aged,Low,0,0,0


In [14]:
print("Risk rows:", len(risk_df))
print("Customer IDs:", len(customer_products))

print(
    "Same number of rows:",
    len(risk_df) == len(customer_products)
)

print(
    "Customer IDs are unique:",
    customer_products["Customer_ID"].nunique() == len(customer_products)
)

print(
    "Customer ID range:",
    customer_products["Customer_ID"].min(),
    "to",
    customer_products["Customer_ID"].max()
)

Risk rows: 29965
Customer IDs: 29965
Same number of rows: True
Customer IDs are unique: True
Customer ID range: 1 to 29965


In [16]:
risk_df["Customer_ID"] = customer_products["Customer_ID"]

print("Customer_ID added successfully!")
print("New shape:", risk_df.shape)

display(
    risk_df[["Customer_ID", "LIMIT_BAL", "AGE"]].head(10)
)

Customer_ID added successfully!
New shape: (29965, 46)


,Customer_ID,LIMIT_BAL,AGE
0,1,20000,24
1,2,120000,26
2,3,90000,34
3,4,50000,37
4,5,50000,57
5,6,50000,37
6,7,500000,29
7,8,100000,23
8,9,140000,28
9,10,20000,35


In [17]:
X_final = risk_df[selected_features].copy()

print("Prediction dataset shape:", X_final.shape)
print("Expected:", (29965, 23))

Prediction dataset shape: (29965, 23)
Expected: (29965, 23)


In [18]:
risk_df["Default_Probability"] = model.predict_proba(X_final)[:, 1]

print("Default probabilities generated successfully!")

display(
    risk_df[
        ["Customer_ID", "Default_Probability"]
    ].head(10)
)

Default probabilities generated successfully!


,Customer_ID,Default_Probability
0,1,0.703450
1,2,0.322857
2,3,0.125238
3,4,0.137028
4,5,0.111693
5,6,0.132175
6,7,0.083133
7,8,0.191453
8,9,0.223781
9,10,0.157205


In [19]:
risk_df["Risk_Prediction"] = (
    risk_df["Default_Probability"] >= FINAL_THRESHOLD
).astype(int)

print("Risk predictions generated successfully!")

display(
    risk_df[
        ["Customer_ID", "Default_Probability", "Risk_Prediction"]
    ].head(10)
)

Risk predictions generated successfully!


,Customer_ID,Default_Probability,Risk_Prediction
0,1,0.703450,1
1,2,0.322857,1
2,3,0.125238,0
3,4,0.137028,0
4,5,0.111693,0
5,6,0.132175,0
6,7,0.083133,0
7,8,0.191453,0
8,9,0.223781,0
9,10,0.157205,0


In [20]:
risk_df["Risk_Category"] = np.where(
    risk_df["Risk_Prediction"] == 1,
    "High Risk",
    "Low Risk"
)

display(
    risk_df[
        ["Customer_ID", "Default_Probability",
         "Risk_Prediction", "Risk_Category"]
    ].head(10)
)

,Customer_ID,Default_Probability,Risk_Prediction,Risk_Category
0,1,0.703450,1,High Risk
1,2,0.322857,1,High Risk
2,3,0.125238,0,Low Risk
3,4,0.137028,0,Low Risk
4,5,0.111693,0,Low Risk
5,6,0.132175,0,Low Risk
6,7,0.083133,0,Low Risk
7,8,0.191453,0,Low Risk
8,9,0.223781,0,Low Risk
9,10,0.157205,0,Low Risk


In [21]:
print("Risk Category Distribution:")
print(risk_df["Risk_Category"].value_counts())

print("\nRisk Category Percentage:")
print(
    (risk_df["Risk_Category"].value_counts(normalize=True) * 100)
    .round(2)
)

Risk Category Distribution:
Risk_Category
Low Risk     22325
High Risk     7640
Name: count, dtype: int64

Risk Category Percentage:
Risk_Category
Low Risk     74.5
High Risk    25.5
Name: proportion, dtype: float64


In [22]:
risk_df["Banking_Decision"] = np.where(
    risk_df["Risk_Category"] == "High Risk",
    "Further Review Required",
    "Standard Evaluation"
)

display(
    risk_df[
        [
            "Customer_ID",
            "Default_Probability",
            "Risk_Category",
            "Banking_Decision"
        ]
    ].head(10)
)

,Customer_ID,Default_Probability,Risk_Category,Banking_Decision
0,1,0.703450,High Risk,Further Review Required
1,2,0.322857,High Risk,Further Review Required
2,3,0.125238,Low Risk,Standard Evaluation
3,4,0.137028,Low Risk,Standard Evaluation
4,5,0.111693,Low Risk,Standard Evaluation
5,6,0.132175,Low Risk,Standard Evaluation
6,7,0.083133,Low Risk,Standard Evaluation
7,8,0.191453,Low Risk,Standard Evaluation
8,9,0.223781,Low Risk,Standard Evaluation
9,10,0.157205,Low Risk,Standard Evaluation


In [23]:
print("Risk Category Distribution:")
print(risk_df["Risk_Category"].value_counts())

print("\nRisk Category Percentage:")
print(
    (risk_df["Risk_Category"].value_counts(normalize=True) * 100)
    .round(2)
)

Risk Category Distribution:
Risk_Category
Low Risk     22325
High Risk     7640
Name: count, dtype: int64

Risk Category Percentage:
Risk_Category
Low Risk     74.5
High Risk    25.5
Name: proportion, dtype: float64


In [24]:
print("Total recommendation records:", len(cross_sell_df))

print(
    "Unique customers with recommendations:",
    cross_sell_df["Customer_ID"].nunique()
)

print(
    "Customers without recommendations:",
    risk_df["Customer_ID"].nunique()
    - cross_sell_df["Customer_ID"].nunique()
)

Total recommendation records: 22190
Unique customers with recommendations: 15398
Customers without recommendations: 14567


In [25]:
recommendation_counts = (
    cross_sell_df
    .groupby("Customer_ID")
    .size()
    .sort_values(ascending=False)
)

print("Recommendation count per customer:")
print(recommendation_counts.value_counts().sort_index())

Recommendation count per customer:
1    9220
2    5564
3     614
Name: count, dtype: int64


In [26]:
cross_sell_summary = (
    cross_sell_df
    .groupby("Customer_ID")
    .agg(
        Recommended_Products=("Recommended_Product", 
                              lambda x: ", ".join(x)),
        Recommendation_Count=("Recommended_Product", "count"),
        Max_Confidence=("Confidence", "max"),
        Max_Lift=("Lift", "max"),
        Max_Support=("Support", "max")
    )
    .reset_index()
)

print("Cross-sell summary created!")
print("Shape:", cross_sell_summary.shape)

display(cross_sell_summary.head(10))

Cross-sell summary created!
Shape: (15398, 6)


,Customer_ID,Recommended_Products,Recommendation_Count,Max_Confidence,Max_Lift,Max_Support
0,1,Credit_Card,1,0.601543,1.267487,0.414715
1,3,"Credit_Card, Savings_Account",2,0.860364,1.134939,0.689419
2,9,"Credit_Card, Debit_Card",2,0.898765,1.171583,0.689419
3,13,Savings_Account,1,0.960844,1.252610,0.414715
4,17,"Insurance, Credit_Card",2,0.695751,1.667993,0.414715
5,18,Credit_Card,1,0.602058,1.268572,0.157713
6,20,Credit_Card,1,0.602058,1.268572,0.157713
7,22,Credit_Card,1,0.601543,1.267487,0.414715
8,23,Credit_Card,1,0.602058,1.268572,0.157713
9,24,Credit_Card,1,0.601543,1.267487,0.414715


In [27]:
final_customer_df = risk_df.merge(
    cross_sell_summary,
    on="Customer_ID",
    how="left"
)

print("Final customer dataset created!")
print("Shape:", final_customer_df.shape)

Final customer dataset created!
Shape: (29965, 55)


In [28]:
final_customer_df["Recommended_Products"] = (
    final_customer_df["Recommended_Products"]
    .fillna("No Recommendation")
)

final_customer_df["Recommendation_Count"] = (
    final_customer_df["Recommendation_Count"]
    .fillna(0)
    .astype(int)
)

final_customer_df["Max_Confidence"] = (
    final_customer_df["Max_Confidence"]
    .fillna(0)
)

final_customer_df["Max_Lift"] = (
    final_customer_df["Max_Lift"]
    .fillna(0)
)

final_customer_df["Max_Support"] = (
    final_customer_df["Max_Support"]
    .fillna(0)
)

In [29]:
display(
    final_customer_df[
        [
            "Customer_ID",
            "Default_Probability",
            "Risk_Category",
            "Banking_Decision",
            "Recommended_Products",
            "Recommendation_Count",
            "Max_Confidence",
            "Max_Lift",
            "Max_Support"
        ]
    ].head(20)
)

,Customer_ID,Default_Probability,Risk_Category,Banking_Decision,Recommended_Products,Recommendation_Count,Max_Confidence,Max_Lift,Max_Support
0,1,0.703450,High Risk,Further Review Required,Credit_Card,1,0.601543,1.267487,0.414715
1,2,0.322857,High Risk,Further Review Required,No Recommendation,0,0.000000,0.000000,0.000000
2,3,0.125238,Low Risk,Standard Evaluation,"Credit_Card, Savings_Account",2,0.860364,1.134939,0.689419
3,4,0.137028,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000
4,5,0.111693,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000
5,6,0.132175,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000
6,7,0.083133,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000
7,8,0.191453,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000
8,9,0.223781,Low Risk,Standard Evaluation,"Credit_Card, Debit_Card",2,0.898765,1.171583,0.689419
9,10,0.157205,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000


In [30]:
def generate_final_action(row):

    if row["Risk_Category"] == "High Risk":
        return "Risk Review Required"

    elif (
        row["Risk_Category"] == "Low Risk"
        and row["Recommendation_Count"] > 0
        and row["Max_Confidence"] >= 0.60
        and row["Max_Lift"] > 1.0
    ):
        return "Eligible for Cross-Sell"

    elif row["Risk_Category"] == "Low Risk":
        return "Standard Banking Service"

    else:
        return "Manual Review"

In [31]:
final_customer_df["Final_Action"] = (
    final_customer_df.apply(generate_final_action, axis=1)
)

print("Final banking actions generated successfully!")

Final banking actions generated successfully!


In [32]:
print("Final Action Distribution:")

print(
    final_customer_df["Final_Action"]
    .value_counts()
)

Final Action Distribution:
Final_Action
Eligible for Cross-Sell     11407
Standard Banking Service    10918
Risk Review Required         7640
Name: count, dtype: int64


In [33]:
display(
    final_customer_df[
        [
            "Customer_ID",
            "Default_Probability",
            "Risk_Category",
            "Banking_Decision",
            "Recommended_Products",
            "Max_Confidence",
            "Max_Lift",
            "Final_Action"
        ]
    ].head(20)
)

,Customer_ID,Default_Probability,Risk_Category,Banking_Decision,Recommended_Products,Max_Confidence,Max_Lift,Final_Action
0,1,0.703450,High Risk,Further Review Required,Credit_Card,0.601543,1.267487,Risk Review Required
1,2,0.322857,High Risk,Further Review Required,No Recommendation,0.000000,0.000000,Risk Review Required
2,3,0.125238,Low Risk,Standard Evaluation,"Credit_Card, Savings_Account",0.860364,1.134939,Eligible for Cross-Sell
3,4,0.137028,Low Risk,Standard Evaluation,No Recommendation,0.000000,0.000000,Standard Banking Service
4,5,0.111693,Low Risk,Standard Evaluation,No Recommendation,0.000000,0.000000,Standard Banking Service
5,6,0.132175,Low Risk,Standard Evaluation,No Recommendation,0.000000,0.000000,Standard Banking Service
6,7,0.083133,Low Risk,Standard Evaluation,No Recommendation,0.000000,0.000000,Standard Banking Service
7,8,0.191453,Low Risk,Standard Evaluation,No Recommendation,0.000000,0.000000,Standard Banking Service
8,9,0.223781,Low Risk,Standard Evaluation,"Credit_Card, Debit_Card",0.898765,1.171583,Eligible for Cross-Sell
9,10,0.157205,Low Risk,Standard Evaluation,No Recommendation,0.000000,0.000000,Standard Banking Service


In [34]:
def customer_profile(customer_id):

    customer = final_customer_df[
        final_customer_df["Customer_ID"] == customer_id
    ]

    if customer.empty:
        print("Customer ID not found.")
        return

    customer = customer.iloc[0]

    print("=" * 55)
    print("          CUSTOMER BANKING PROFILE")
    print("=" * 55)

    print(f"Customer ID          : {customer['Customer_ID']}")
    print()

    print("CREDIT RISK")
    print("-" * 55)
    print(
        f"Default Probability  : "
        f"{customer['Default_Probability']:.2%}"
    )
    print(f"Risk Category        : {customer['Risk_Category']}")
    print(f"Banking Decision     : {customer['Banking_Decision']}")
    print()

    print("CROSS-SELL")
    print("-" * 55)
    print(
        f"Recommended Products: "
        f"{customer['Recommended_Products']}"
    )
    print(
        f"Confidence           : "
        f"{customer['Max_Confidence']:.2%}"
    )
    print(
        f"Lift                 : "
        f"{customer['Max_Lift']:.2f}"
    )
    print()

    print("FINAL ACTION")
    print("-" * 55)
    print(f"{customer['Final_Action']}")

    print("=" * 55)

In [35]:
customer_profile(1)

          CUSTOMER BANKING PROFILE
Customer ID          : 1

CREDIT RISK
-------------------------------------------------------
Default Probability  : 70.34%
Risk Category        : High Risk
Banking Decision     : Further Review Required

CROSS-SELL
-------------------------------------------------------
Recommended Products: Credit_Card
Confidence           : 60.15%
Lift                 : 1.27

FINAL ACTION
-------------------------------------------------------
Risk Review Required


In [36]:
customer_profile(3)

          CUSTOMER BANKING PROFILE
Customer ID          : 3

CREDIT RISK
-------------------------------------------------------
Default Probability  : 12.52%
Risk Category        : Low Risk
Banking Decision     : Standard Evaluation

CROSS-SELL
-------------------------------------------------------
Recommended Products: Credit_Card, Savings_Account
Confidence           : 86.04%
Lift                 : 1.13

FINAL ACTION
-------------------------------------------------------
Eligible for Cross-Sell


In [37]:
customer_profile(17)

          CUSTOMER BANKING PROFILE
Customer ID          : 17

CREDIT RISK
-------------------------------------------------------
Default Probability  : 53.87%
Risk Category        : High Risk
Banking Decision     : Further Review Required

CROSS-SELL
-------------------------------------------------------
Recommended Products: Insurance, Credit_Card
Confidence           : 69.58%
Lift                 : 1.67

FINAL ACTION
-------------------------------------------------------
Risk Review Required


In [40]:
def risk_explanation(customer_id):

    customer = final_customer_df[
        final_customer_df["Customer_ID"] == customer_id
    ]

    if customer.empty:
        print("Customer ID not found.")
        return

    customer = customer.iloc[0]

    print("=" * 60)
    print("             RISK EXPLANATION")
    print("=" * 60)

    print(f"Customer ID: {customer_id}")
    print(f"Default Probability: {customer['Default_Probability']:.2%}")
    print(f"Risk Category: {customer['Risk_Category']}")
    print()

    print("KEY CUSTOMER RISK INDICATORS")
    print("-" * 60)

    print(f"Recent Payment Delay : {customer['Recent_Payment_Delay']}")
    print(f"Recent Average Delay : {customer['Recent_Avg_Delay']:.2f}")
    print(f"Max Payment Delay    : {customer['Max_Payment_Delay']}")
    print(f"Severe Delay Count   : {customer['Severe_Delay_Count']}")
    print(f"Frequent Delay Flag  : {customer['Frequent_Delay_Flag']}")
    print(f"Severe Delay Flag    : {customer['Severe_Delay_Flag']}")

    print("=" * 60)

In [39]:
print([
    "Recent_Payment_Delay",
    "Recent_Avg_Delay",
    "Max_Payment_Delay",
    "Severe_Delay_Count",
    "Frequent_Delay_Flag",
    "Severe_Delay_Flag"
])

print(
    final_customer_df[
        [
            "Recent_Payment_Delay",
            "Recent_Avg_Delay",
            "Max_Payment_Delay",
            "Severe_Delay_Count",
            "Frequent_Delay_Flag",
            "Severe_Delay_Flag"
        ]
    ].head()
)

['Recent_Payment_Delay', 'Recent_Avg_Delay', 'Max_Payment_Delay', 'Severe_Delay_Count', 'Frequent_Delay_Flag', 'Severe_Delay_Flag']
   Recent_Payment_Delay  Recent_Avg_Delay  Max_Payment_Delay  \
0                     2          1.000000                  2   
1                    -1          0.333333                  2   
2                     0          0.000000                  0   
3                     0          0.000000                  0   
4                    -1         -0.666667                  0   

   Severe_Delay_Count  Frequent_Delay_Flag  Severe_Delay_Flag  
0                   2                    0                  1  
1                   2                    0                  1  
2                   0                    0                  0  
3                   0                    0                  0  
4                   0                    0                  0  


In [41]:
def risk_explanation(customer_id):

    customer = final_customer_df[
        final_customer_df["Customer_ID"] == customer_id
    ]

    if customer.empty:
        print("Customer ID not found.")
        return

    customer = customer.iloc[0]

    print("=" * 60)
    print("             RISK EXPLANATION")
    print("=" * 60)

    print(f"Customer ID          : {customer_id}")
    print(f"Default Probability  : {customer['Default_Probability']:.2%}")
    print(f"Risk Category        : {customer['Risk_Category']}")
    print()

    print("KEY RISK INDICATORS")
    print("-" * 60)

    print(f"Recent Payment Delay : {customer['Recent_Payment_Delay']}")
    print(f"Recent Average Delay : {customer['Recent_Avg_Delay']:.2f}")
    print(f"Maximum Payment Delay: {customer['Max_Payment_Delay']}")
    print(f"Severe Delay Count   : {customer['Severe_Delay_Count']}")
    print(f"Frequent Delay Flag  : {customer['Frequent_Delay_Flag']}")
    print(f"Severe Delay Flag    : {customer['Severe_Delay_Flag']}")

    print()
    print("INTERPRETATION")
    print("-" * 60)

    reasons = []

    if customer["Recent_Payment_Delay"] > 0:
        reasons.append("Recent payment delay is present.")

    if customer["Recent_Avg_Delay"] > 0:
        reasons.append("Average recent payment delay is positive.")

    if customer["Max_Payment_Delay"] > 0:
        reasons.append("Customer has experienced payment delays.")

    if customer["Severe_Delay_Count"] > 0:
        reasons.append(
            f"{int(customer['Severe_Delay_Count'])} severe payment delay(s) detected."
        )

    if customer["Frequent_Delay_Flag"] == 1:
        reasons.append("Frequent payment delays detected.")

    if customer["Severe_Delay_Flag"] == 1:
        reasons.append("Severe payment behavior detected.")

    if len(reasons) == 0:
        reasons.append("No major engineered payment-delay indicators detected.")

    for i, reason in enumerate(reasons, 1):
        print(f"{i}. {reason}")

    print("=" * 60)

In [42]:
risk_explanation(17)

             RISK EXPLANATION
Customer ID          : 17
Default Probability  : 53.87%
Risk Category        : High Risk

KEY RISK INDICATORS
------------------------------------------------------------
Recent Payment Delay : 0
Recent Average Delay : 0.67
Maximum Payment Delay: 2
Severe Delay Count   : 4
Frequent Delay Flag  : 1
Severe Delay Flag    : 1

INTERPRETATION
------------------------------------------------------------
1. Average recent payment delay is positive.
2. Customer has experienced payment delays.
3. 4 severe payment delay(s) detected.
4. Frequent payment delays detected.
5. Severe payment behavior detected.


In [43]:
final_output = final_customer_df[
    [
        "Customer_ID",
        "Default_Probability",
        "Risk_Category",
        "Banking_Decision",
        "Recommended_Products",
        "Recommendation_Count",
        "Max_Confidence",
        "Max_Lift",
        "Max_Support",
        "Final_Action"
    ]
].copy()

print("Final output created successfully!")
print("Shape:", final_output.shape)

display(final_output.head(10))

Final output created successfully!
Shape: (29965, 10)


,Customer_ID,Default_Probability,Risk_Category,Banking_Decision,Recommended_Products,Recommendation_Count,Max_Confidence,Max_Lift,Max_Support,Final_Action
0,1,0.703450,High Risk,Further Review Required,Credit_Card,1,0.601543,1.267487,0.414715,Risk Review Required
1,2,0.322857,High Risk,Further Review Required,No Recommendation,0,0.000000,0.000000,0.000000,Risk Review Required
2,3,0.125238,Low Risk,Standard Evaluation,"Credit_Card, Savings_Account",2,0.860364,1.134939,0.689419,Eligible for Cross-Sell
3,4,0.137028,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
4,5,0.111693,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
5,6,0.132175,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
6,7,0.083133,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
7,8,0.191453,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
8,9,0.223781,Low Risk,Standard Evaluation,"Credit_Card, Debit_Card",2,0.898765,1.171583,0.689419,Eligible for Cross-Sell
9,10,0.157205,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service


In [44]:
import importlib.util

if importlib.util.find_spec("streamlit"):
    print("Streamlit is installed!")
else:
    print("Streamlit is NOT installed.")

Streamlit is NOT installed.


In [45]:
!pip install streamlit

  Using cached click-8.4.2-py3-none-any.whl.metadata (2.6 kB)
  Using cached tenacity-9.1.4-py3-none-any.whl.metadata (1.2 kB)
   ---------------------------------------- 0.0/10.5 MB ? eta -:--:--
   ---------------------------------------- 0.0/10.5 MB ? eta -:--:--
    --------------------------------------- 0.3/10.5 MB ? eta -:--:--
    --------------------------------------- 0.3/10.5 MB ? eta -:--:--
    --------------------------------------- 0.3/10.5 MB ? eta -:--:--
   - -------------------------------------- 0.5/10.5 MB 508.0 kB/s eta 0:00:20
   - -------------------------------------- 0.5/10.5 MB 508.0 kB/s eta 0:00:20
   - -------------------------------------- 0.5/10.5 MB 508.0 kB/s eta 0:00:20
   - -------------------------------------- 0.5/10.5 MB 508.0 kB/s eta 0:00:20
   -- ------------------------------------- 0.8/10.5 MB 376.9 kB/s eta 0:00:26
   -- ------------------------------------- 0.8/10.5 MB 376.9 kB/s eta 0:00:26
   -- ------------------------------------- 0.8/1

In [46]:
import importlib.util

if importlib.util.find_spec("streamlit"):
    print("Streamlit is installed!")
else:
    print("Streamlit is NOT installed.")

Streamlit is installed!


In [47]:
import pandas as pd
import numpy as np
import joblib
import os

print("Phase 10 - Final Banking System")
print("Current folder:", os.getcwd())


Phase 10 - Final Banking System
Current folder: c:\Users\Meghana\OneDrive\Desktop\BANKING-CREDIT-RISK\notebooks


In [48]:
import joblib

final_model = joblib.load("final_credit_risk_model.pkl")
final_features = joblib.load("final_features.pkl")

print("Final model loaded successfully!")
print("Number of features:", len(final_features))
print("Final features:")
print(final_features)


Final model loaded successfully!
Number of features: 23
Final features:
['LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_SEPT', 'PAY_AUG', 'PAY_JUL', 'PAY_JUN', 'PAY_MAY', 'PAY_APR', 'BILL_AMT_SEPT', 'BILL_AMT_AUG', 'BILL_AMT_JUL', 'BILL_AMT_JUN', 'BILL_AMT_MAY', 'BILL_AMT_APR', 'PAY_AMT_SEPT', 'PAY_AMT_AUG', 'PAY_AMT_JUL', 'PAY_AMT_JUN', 'PAY_AMT_MAY', 'PAY_AMT_APR']


In [49]:
cross_sell_df = pd.read_csv(
    "customer_cross_sell_recommendations.csv"
)

print("Cross-sell dataset loaded successfully!")
print("Shape:", cross_sell_df.shape)
print("Columns:", cross_sell_df.columns.tolist())

display(cross_sell_df.head())

Cross-sell dataset loaded successfully!
Shape: (22190, 6)
Columns: ['Customer_ID', 'Existing_Products', 'Recommended_Product', 'Confidence', 'Lift', 'Support']


,Customer_ID,Existing_Products,Recommended_Product,Confidence,Lift,Support
0,1,"Savings_Account, Debit_Card",Credit_Card,0.601543,1.267487,0.414715
1,3,Debit_Card,Credit_Card,0.538636,1.134939,0.431615
2,3,Debit_Card,Savings_Account,0.860364,1.121618,0.689419
3,9,"Savings_Account, Insurance",Credit_Card,0.556027,1.171583,0.162302
4,9,"Savings_Account, Insurance",Debit_Card,0.898765,1.121618,0.689419


In [50]:
risk_df = pd.read_csv(
    "credit_default_feature_engineered.csv"
)

print("Risk dataset loaded successfully!")
print("Shape:", risk_df.shape)
print("Number of columns:", len(risk_df.columns))

display(risk_df.head())

Risk dataset loaded successfully!
Shape: (29965, 45)
Number of columns: 45


,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_SEPT,PAY_AUG,PAY_JUL,PAY_JUN,PAY_MAY,...,Severe_Delay_Count,Recent_Payment_Delay,Recent_Avg_Delay,Older_Avg_Delay,Payment_Delay_Trend,Age_Category,Credit_Limit_Category,High_Utilization_Flag,Frequent_Delay_Flag,Severe_Delay_Flag
0,20000,2,2,1,24,2,2,-1,-1,-2,...,2,2,1.000000,-1.666667,2.666667,Young,Low,0,0,1
1,120000,2,2,2,26,-1,2,0,0,0,...,2,-1,0.333333,0.666667,-0.333333,Adult,Medium,0,0,1
2,90000,2,2,2,34,0,0,0,0,0,...,0,0,0.000000,0.000000,0.000000,Adult,Medium,0,0,0
3,50000,2,2,1,37,0,0,0,0,0,...,0,0,0.000000,0.000000,0.000000,Adult,Low,1,0,0
4,50000,1,2,1,57,-1,0,-1,0,0,...,0,-1,-0.666667,0.000000,-0.666667,Middle_Aged,Low,0,0,0


In [51]:
if "Customer_ID" not in risk_df.columns:
    risk_df.insert(
        0,
        "Customer_ID",
        range(1, len(risk_df) + 1)
    )

print("Customer ID added successfully!")
print("Number of customers:", risk_df["Customer_ID"].nunique())
print("ID range:",
      risk_df["Customer_ID"].min(),
      "to",
      risk_df["Customer_ID"].max())

display(
    risk_df[["Customer_ID", "LIMIT_BAL", "AGE"]].head(10)
)

Customer ID added successfully!
Number of customers: 29965
ID range: 1 to 29965


,Customer_ID,LIMIT_BAL,AGE
0,1,20000,24
1,2,120000,26
2,3,90000,34
3,4,50000,37
4,5,50000,57
5,6,50000,37
6,7,500000,29
7,8,100000,23
8,9,140000,28
9,10,20000,35


In [52]:
X_final = risk_df[final_features]

print("Input data shape:", X_final.shape)
print("Expected features:", len(final_features))

risk_df["Default_Probability"] = (
    final_model.predict_proba(X_final)[:, 1]
)

print("Default probabilities generated successfully!")

display(
    risk_df[
        ["Customer_ID", "Default_Probability"]
    ].head(10)
)

Input data shape: (29965, 23)
Expected features: 23
Default probabilities generated successfully!


,Customer_ID,Default_Probability
0,1,0.703450
1,2,0.322857
2,3,0.125238
3,4,0.137028
4,5,0.111693
5,6,0.132175
6,7,0.083133
7,8,0.191453
8,9,0.223781
9,10,0.157205


In [53]:
FINAL_THRESHOLD = 0.25

risk_df["Risk_Category"] = np.where(
    risk_df["Default_Probability"] >= FINAL_THRESHOLD,
    "High Risk",
    "Low Risk"
)

print("Risk categories created successfully!")

print("\nRisk Category Distribution:")
print(risk_df["Risk_Category"].value_counts())

print("\nRisk Category Percentage:")
print(
    risk_df["Risk_Category"]
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)

Risk categories created successfully!

Risk Category Distribution:
Risk_Category
Low Risk     22325
High Risk     7640
Name: count, dtype: int64

Risk Category Percentage:
Risk_Category
Low Risk     74.5
High Risk    25.5
Name: proportion, dtype: float64


In [54]:
risk_df["Banking_Decision"] = np.where(
    risk_df["Risk_Category"] == "High Risk",
    "Further Review Required",
    "Standard Evaluation"
)

print("Banking decisions created successfully!")

print("\nBanking Decision Distribution:")
print(risk_df["Banking_Decision"].value_counts())

display(
    risk_df[
        [
            "Customer_ID",
            "Default_Probability",
            "Risk_Category",
            "Banking_Decision"
        ]
    ].head(10)
)



Banking decisions created successfully!

Banking Decision Distribution:
Banking_Decision
Standard Evaluation        22325
Further Review Required     7640
Name: count, dtype: int64


,Customer_ID,Default_Probability,Risk_Category,Banking_Decision
0,1,0.703450,High Risk,Further Review Required
1,2,0.322857,High Risk,Further Review Required
2,3,0.125238,Low Risk,Standard Evaluation
3,4,0.137028,Low Risk,Standard Evaluation
4,5,0.111693,Low Risk,Standard Evaluation
5,6,0.132175,Low Risk,Standard Evaluation
6,7,0.083133,Low Risk,Standard Evaluation
7,8,0.191453,Low Risk,Standard Evaluation
8,9,0.223781,Low Risk,Standard Evaluation
9,10,0.157205,Low Risk,Standard Evaluation


In [55]:
cross_sell_summary = (
    cross_sell_df
    .groupby("Customer_ID")
    .agg(
        Recommended_Products=(
            "Recommended_Product",
            lambda x: ", ".join(x)
        ),
        Recommendation_Count=(
            "Recommended_Product",
            "count"
        ),
        Max_Confidence=(
            "Confidence",
            "max"
        ),
        Max_Lift=(
            "Lift",
            "max"
        ),
        Max_Support=(
            "Support",
            "max"
        )
    )
    .reset_index()
)

print("Cross-sell summary created successfully!")
print("Shape:", cross_sell_summary.shape)

display(cross_sell_summary.head(10))

Cross-sell summary created successfully!
Shape: (15398, 6)


,Customer_ID,Recommended_Products,Recommendation_Count,Max_Confidence,Max_Lift,Max_Support
0,1,Credit_Card,1,0.601543,1.267487,0.414715
1,3,"Credit_Card, Savings_Account",2,0.860364,1.134939,0.689419
2,9,"Credit_Card, Debit_Card",2,0.898765,1.171583,0.689419
3,13,Savings_Account,1,0.960844,1.252610,0.414715
4,17,"Insurance, Credit_Card",2,0.695751,1.667993,0.414715
5,18,Credit_Card,1,0.602058,1.268572,0.157713
6,20,Credit_Card,1,0.602058,1.268572,0.157713
7,22,Credit_Card,1,0.601543,1.267487,0.414715
8,23,Credit_Card,1,0.602058,1.268572,0.157713
9,24,Credit_Card,1,0.601543,1.267487,0.414715


In [58]:
print("DataFrames currently available:\n")

for name in list(globals().keys()):
    obj = globals()[name]
    if hasattr(obj, "shape") and hasattr(obj, "columns"):
        print(f"{name} → {obj.shape}")

DataFrames currently available:

cross_sell_df → (22190, 6)
customer_products → (29965, 7)
customer_products_v2 → (29965, 7)
risk_df → (29965, 49)
X_final → (29965, 23)
cross_sell_summary → (15398, 6)
final_customer_df → (29965, 56)
final_output → (29965, 10)


In [59]:
print("Shape:", final_output.shape)

print("\nColumns:")
print(final_output.columns.tolist())

print("\nFirst 10 rows:")
display(final_output.head(10))

Shape: (29965, 10)

Columns:
['Customer_ID', 'Default_Probability', 'Risk_Category', 'Banking_Decision', 'Recommended_Products', 'Recommendation_Count', 'Max_Confidence', 'Max_Lift', 'Max_Support', 'Final_Action']

First 10 rows:


,Customer_ID,Default_Probability,Risk_Category,Banking_Decision,Recommended_Products,Recommendation_Count,Max_Confidence,Max_Lift,Max_Support,Final_Action
0,1,0.703450,High Risk,Further Review Required,Credit_Card,1,0.601543,1.267487,0.414715,Risk Review Required
1,2,0.322857,High Risk,Further Review Required,No Recommendation,0,0.000000,0.000000,0.000000,Risk Review Required
2,3,0.125238,Low Risk,Standard Evaluation,"Credit_Card, Savings_Account",2,0.860364,1.134939,0.689419,Eligible for Cross-Sell
3,4,0.137028,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
4,5,0.111693,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
5,6,0.132175,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
6,7,0.083133,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
7,8,0.191453,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
8,9,0.223781,Low Risk,Standard Evaluation,"Credit_Card, Debit_Card",2,0.898765,1.171583,0.689419,Eligible for Cross-Sell
9,10,0.157205,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service


In [60]:
print("Missing values:")
print(final_output.isnull().sum())

print("\nDuplicate rows:", final_output.duplicated().sum())

print("\nNumber of customers:", final_output["Customer_ID"].nunique())

Missing values:
Customer_ID             0
Default_Probability     0
Risk_Category           0
Banking_Decision        0
Recommended_Products    0
Recommendation_Count    0
Max_Confidence          0
Max_Lift                0
Max_Support             0
Final_Action            0
dtype: int64

Duplicate rows: 0

Number of customers: 29965


In [61]:
print("Risk Category Distribution:")
print(final_output["Risk_Category"].value_counts())

print("\nBanking Decision Distribution:")
print(final_output["Banking_Decision"].value_counts())

print("\nFinal Action Distribution:")
print(final_output["Final_Action"].value_counts())

Risk Category Distribution:
Risk_Category
Low Risk     22325
High Risk     7640
Name: count, dtype: int64

Banking Decision Distribution:
Banking_Decision
Standard Evaluation        22325
Further Review Required     7640
Name: count, dtype: int64

Final Action Distribution:
Final_Action
Eligible for Cross-Sell     11407
Standard Banking Service    10918
Risk Review Required         7640
Name: count, dtype: int64


In [62]:
print("Recommendation Count Distribution:")
print(final_output["Recommendation_Count"].value_counts().sort_index())

print("\nTotal customers with recommendations:",
      (final_output["Recommendation_Count"] > 0).sum())

print("Total customers without recommendations:",
      (final_output["Recommendation_Count"] == 0).sum())

print("\nMaximum recommendations for one customer:",
      final_output["Recommendation_Count"].max())

Recommendation Count Distribution:
Recommendation_Count
0    14567
1     9220
2     5564
3      614
Name: count, dtype: int64

Total customers with recommendations: 15398
Total customers without recommendations: 14567

Maximum recommendations for one customer: 3


In [63]:
print(
    pd.crosstab(
        final_output["Risk_Category"],
        final_output["Final_Action"]
    )
)

Final_Action   Eligible for Cross-Sell  Risk Review Required  \
Risk_Category                                                  
High Risk                            0                  7640   
Low Risk                         11407                     0   

Final_Action   Standard Banking Service  
Risk_Category                            
High Risk                             0  
Low Risk                          10918  


In [64]:
recommended = final_output[
    final_output["Recommendation_Count"] > 0
]

print("Customers with recommendations:", len(recommended))

print("\nConfidence statistics:")
print(recommended["Max_Confidence"].describe())

print("\nLift statistics:")
print(recommended["Max_Lift"].describe())

print("\nSupport statistics:")
print(recommended["Max_Support"].describe())

Customers with recommendations: 15398

Confidence statistics:
count    15398.000000
mean         0.732736
std          0.135305
min          0.601543
25%          0.601543
50%          0.695751
75%          0.864021
max          0.973745
Name: Max_Confidence, dtype: float64

Lift statistics:
count    15398.000000
mean         1.317436
std          0.190774
min          1.134939
25%          1.171184
50%          1.267487
75%          1.268572
max          1.678660
Name: Max_Lift, dtype: float64

Support statistics:
count    15398.000000
mean         0.380698
std          0.195779
min          0.102422
25%          0.171262
50%          0.414715
75%          0.431615
max          0.689419
Name: Max_Support, dtype: float64


In [65]:
print("========== PHASE 10 INTEGRATION SUMMARY ==========")

print("\nTotal Customers:", len(final_output))

print("\nRisk Distribution:")
print(final_output["Risk_Category"].value_counts())

print("\nFinal Action Distribution:")
print(final_output["Final_Action"].value_counts())

print("\nCross-Sell Summary:")
print("Customers with recommendations:",
      (final_output["Recommendation_Count"] > 0).sum())
print("Customers without recommendations:",
      (final_output["Recommendation_Count"] == 0).sum())

print("\nRecommendation Count:")
print(final_output["Recommendation_Count"].value_counts().sort_index())

========== PHASE 10 INTEGRATION SUMMARY ==========

Total Customers: 29965

Risk Distribution:
Risk_Category
Low Risk     22325
High Risk     7640
Name: count, dtype: int64

Final Action Distribution:
Final_Action
Eligible for Cross-Sell     11407
Standard Banking Service    10918
Risk Review Required         7640
Name: count, dtype: int64

Cross-Sell Summary:
Customers with recommendations: 15398
Customers without recommendations: 14567

Recommendation Count:
Recommendation_Count
0    14567
1     9220
2     5564
3      614
Name: count, dtype: int64


In [66]:
final_output.to_csv(
    "Phase_10_Final_Integrated_Banking_Output.csv",
    index=False
)

print("File saved successfully!")

print("\nRows:", final_output.shape[0])
print("Columns:", final_output.shape[1])

print("\nSaved columns:")
print(final_output.columns.tolist())

File saved successfully!

Rows: 29965
Columns: 10

Saved columns:
['Customer_ID', 'Default_Probability', 'Risk_Category', 'Banking_Decision', 'Recommended_Products', 'Recommendation_Count', 'Max_Confidence', 'Max_Lift', 'Max_Support', 'Final_Action']


In [67]:
import pandas as pd

check_df = pd.read_csv("Phase_10_Final_Integrated_Banking_Output.csv")

print("File loaded successfully!")

print("\nShape:", check_df.shape)

print("\nMissing values:")
print(check_df.isnull().sum().sum())

print("\nDuplicate rows:", check_df.duplicated().sum())

print("\nFirst 5 rows:")
display(check_df.head())

File loaded successfully!

Shape: (29965, 10)

Missing values:
0

Duplicate rows: 0

First 5 rows:


,Customer_ID,Default_Probability,Risk_Category,Banking_Decision,Recommended_Products,Recommendation_Count,Max_Confidence,Max_Lift,Max_Support,Final_Action
0,1,0.703450,High Risk,Further Review Required,Credit_Card,1,0.601543,1.267487,0.414715,Risk Review Required
1,2,0.322857,High Risk,Further Review Required,No Recommendation,0,0.000000,0.000000,0.000000,Risk Review Required
2,3,0.125238,Low Risk,Standard Evaluation,"Credit_Card, Savings_Account",2,0.860364,1.134939,0.689419,Eligible for Cross-Sell
3,4,0.137028,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service
4,5,0.111693,Low Risk,Standard Evaluation,No Recommendation,0,0.000000,0.000000,0.000000,Standard Banking Service


In [68]:
import os

print(os.path.abspath("Phase_10_Final_Integrated_Banking_Output.csv"))

c:\Users\Meghana\OneDrive\Desktop\BANKING-CREDIT-RISK\notebooks\Phase_10_Final_Integrated_Banking_Output.csv


In [69]:
import os

print(os.listdir())

['01_dataset_check.ipynb', '01_eda.ipynb', '02_preprocessing.ipynb', '03_model_training.ipynb', '04_association_rules.ipynb', '05_final_banking_system.ipynb', 'credit_default_feature_engineered.csv', 'customer_cross_sell_recommendations.csv', 'customer_products.csv', 'customer_products_v2.csv', 'final_credit_risk_model.pkl', 'final_features.pkl', 'Phase_10_Final_Integrated_Banking_Output.csv']


In [70]:
import os

print(os.listdir())

['01_dataset_check.ipynb', '01_eda.ipynb', '02_preprocessing.ipynb', '03_model_training.ipynb', '04_association_rules.ipynb', '05_final_banking_system.ipynb', 'credit_default_feature_engineered.csv', 'customer_cross_sell_recommendations.csv', 'customer_products.csv', 'customer_products_v2.csv', 'final_credit_risk_model.pkl', 'final_features.pkl', 'Phase_10_Final_Integrated_Banking_Output.csv']


# 🔵 Phase 10 — Final Integration & Banking Decision System

## Objective

The objective of Phase 10 was to integrate the trained credit-risk prediction model with the Phase 9 cross-sell recommendation system to create a unified customer-level banking decision system.

The integrated system combines:

- Credit default probability prediction
- Risk classification using the optimized threshold
- Banking decision generation
- Personalized cross-sell recommendations
- Recommendation confidence, lift, and support
- Final customer-level banking action

---

## 1. Final Risk Prediction

The saved final machine learning model was used to generate default probabilities for all **29,965 customers**.

The optimized classification threshold obtained during Phase 7 was:

**Final Threshold = 0.25**

Customers with a predicted default probability greater than or equal to 0.25 were classified as **High Risk**, while the remaining customers were classified as **Low Risk**.

### Risk Distribution

| Risk Category | Customers | Percentage |
|---|---:|---:|
| Low Risk | 22,325 | 74.50% |
| High Risk | 7,640 | 25.50% |
| **Total** | **29,965** | **100%** |

---

## 2. Banking Decision Generation

Risk categories were converted into business-oriented banking decisions.

| Risk Category | Banking Decision |
|---|---|
| High Risk | Further Review Required |
| Low Risk | Standard Evaluation |

Therefore, all high-risk customers are directed for additional review, while low-risk customers proceed with standard banking evaluation.

---

## 3. Cross-Sell Recommendation Integration

The association-rule-based recommendation system developed in Phase 9 was integrated with the customer risk information.

For each customer, the system stores:

- Recommended products
- Number of recommended products
- Maximum confidence
- Maximum lift
- Maximum support

A total of **15,398 customers** received at least one product recommendation, while **14,567 customers** had no recommendation.

### Recommendation Distribution

| Recommendation Count | Customers |
|---:|---:|
| 0 | 14,567 |
| 1 | 9,220 |
| 2 | 5,564 |
| 3 | 614 |
| **Total** | **29,965** |

The maximum number of recommended products for a single customer was **3**.

---

## 4. Recommendation Quality

For customers receiving recommendations, the integrated recommendation metrics were validated.

| Metric | Mean | Minimum | Maximum |
|---|---:|---:|---:|
| Confidence | 0.7327 | 0.6015 | 0.9737 |
| Lift | 1.3174 | 1.1349 | 1.6787 |

The observed lift values were greater than 1, indicating that the recommended product associations were stronger than random co-occurrence within the analyzed data.

---

## 5. Final Banking Action

The final business action was generated by combining customer risk and cross-sell eligibility.

| Final Action | Customers |
|---|---:|
| Eligible for Cross-Sell | 11,407 |
| Standard Banking Service | 10,918 |
| Risk Review Required | 7,640 |
| **Total** | **29,965** |

The business logic was validated using a cross-tabulation of risk category and final action.

- High-risk customers were assigned **Risk Review Required**.
- Low-risk customers with recommendations were assigned **Eligible for Cross-Sell**.
- Low-risk customers without recommendations were assigned **Standard Banking Service**.

---

## 6. Final Integrated Dataset

The final integrated dataset contains **29,965 rows and 10 columns**.

### Final Columns

1. `Customer_ID`
2. `Default_Probability`
3. `Risk_Category`
4. `Banking_Decision`
5. `Recommended_Products`
6. `Recommendation_Count`
7. `Max_Confidence`
8. `Max_Lift`
9. `Max_Support`
10. `Final_Action`

The final dataset was validated successfully:

- Missing values: **0**
- Duplicate rows: **0**
- Unique customers: **29,965**

---

## 7. Final Output File

The completed Phase 10 integrated dataset was saved as:

`Phase_10_Final_Integrated_Banking_Output.csv`

The saved CSV was successfully loaded again and verified.

---

## 8. Overall System Flow

The completed banking decision system follows the workflow:

**Customer Data**  
↓  
**Feature Engineering & Final Features**  
↓  
**Credit Risk Prediction**  
↓  
**Default Probability**  
↓  
**Threshold = 0.25**  
↓  
**Risk Category**  
↓  
**Banking Decision**  
↓  
**Cross-Sell Recommendation System**  
↓  
**Final Banking Action**

---

## Conclusion

Phase 10 successfully integrated the credit-risk prediction system and the cross-sell recommendation system into a unified customer-level banking decision framework.

The final system can identify customer risk, generate appropriate banking decisions, and provide personalized product recommendations for eligible low-risk customers. This integration transforms the individual machine-learning components developed in earlier phases into a complete banking decision-support system.